ENH: decoding and ICC color transform on webworker

Punzo opened the pull request:
Implements https://github.com/herrmannlab/slim/issues/94 

See https://github.com/herrmannlab/slim/pull/93 for more details

hackermd reviewed (changes requested):
@Punzo this is awesome! Thanks so much.

I haven't tested the code yet, but have a few general comments and suggestions. I really like the fact that you have factored out and abstracted the individual decoders and the transformer. I think there would be opportunities to further abstract it and define a constant interface for decoders and transformers by deriving them from a common abstract base class (see individual comments for details).

In addition, I am wondering whether we could improve the API for the web workers and expose some functionality at the level of the main viewer class. For example, it would be nice if the creation and termination of worker processes/threads could be controlled (and potentially parametrized) via the `render()` and `cleanup()` methods.

Punzo commented on package.json:
```diff
@@ -47,17 +48,47 @@
+    "webpack-bundle-analyzer": "^4.4.2",
+    "webpack-cli": "^4.8.0",
+    "webpack-dev-server": "4.1.0",
+    "worker-loader": "^3.0.8"
```
thanks for point out, I removed all teh uncessary libraries in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/2ca9fa303c15811fb095e22d3ade7e4994914c76

Punzo commented on src/decode.js:
```diff
@@ -1,222 +1,112 @@
   switch (bitsAllocated) {
     case 1:
-      pixelArray = dcmjs.data.BitArray.unpack(transformedFrame) // Uint8Array
+      pixelArray = dcmjs.data.BitArray.unpack(result.pixelData) // Uint8Array
```
renamed frameData in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/d4769104fdb5117b0f1aad42573fb4ded97941d9

Punzo commented on src/pyramid.js:
```diff
@@ -1,9 +1,48 @@
+ */
+function _getICCProfiles(pyramid, client) {
+  const metadata = pyramid.metadata
+  const ICCProfiles = []
```
renamed in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/45f6461792a31ca39e1ace3537cfeeccb39a9106

Punzo commented on src/viewer.js:
```diff
@@ -1065,7 +958,8 @@ class VolumeImageViewer {
             client: this[_options].client,
-            channel: opticalPathIdentifier
+            channel: opticalPathIdentifier,
+            ICCProfiles: _getICCProfiles(pyramid, this[_options].client)
```
renamed in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/45f6461792a31ca39e1ace3537cfeeccb39a9106

Punzo commented on src/webWorker/decodeAndTrasformTask.js:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    // Apply ICC color transform
+    transformICCAsync(data.ICCProfiles, data.sopInstanceUID, decodedFrame).then((transformedFrame) => {
```
yes, there is a check in transformAsync. If  the iccProfiles array is undefined or it has zero elements, the function just return the original decoded frame without appliying the transfrom. I have updated the function description in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/a9e171eee921815b830d9af413fefeaf531d5489

Punzo commented on src/webWorker/decodeAndTrasformTask.js:
```diff
@@ -0,0 +1,178 @@
+function handler(data, doneCallback) {
+  _checkImageTypeAndDecode(
+    data
+  ).then(({data, decodedFrame, frameInfo}) => {
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/f79b013c3f6098e20a9fa81b5d23409dace37eb9

Punzo commented on src/webWorker/decodeAndTrasformTask.js:
```diff
@@ -0,0 +1,178 @@
+ */
+async function _checkImageTypeAndDecode ({
+  data
+}) {
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/f79b013c3f6098e20a9fa81b5d23409dace37eb9

Punzo commented on src/webWorker/decoders/decodeJPEGBaseline8Bit.js:
```diff
@@ -0,0 +1,56 @@
+ * @returns {object} decoded array and frameInfo
+ * @private
+ */
+async function decodeAsync(byteArray) {
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/899dc0c605e0970b6ee1f794e14bfbcd66367435

Punzo commented on src/webWorker/transformers/transformICC.js:
```diff
@@ -0,0 +1,78 @@
+ * @returns {Buffer} transformed buffer
+ * @private
+ */
+ async function transformAsync(ICCProfiles, sopInstanceUID, decodedFrame) {
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/fb062d379c27700fadfe748e45f8fce2b9c6adcd

Punzo commented on src/webWorker/webWorkerManager.js:
```diff
@@ -0,0 +1,319 @@
+    return
+  }
+
+  const worker = new dicomMicroscopyImageLoaderWebWorker();
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/537c2059adc4eae8435288f0bcbcd52c755cce3a

Punzo commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,140 @@
+  config = data.config
+
+  // Additional web worker tasks can self-register by calling self.registerTaskHandler
+  self.registerTaskHandler = registerTaskHandler
```
no self here is the webWorker. Essentially the worker loader load the file index.worker.js and it creates an object WebWorker with the file as input in the constructor. So the self points to the worker context which is also completly separate by the main threads. Also for constructions needs a pretty specific syntax.

Punzo commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,140 @@
+ * Initialization function that loads additional web workers and initializes them
+ * @param data
+ */
+function initialize(data) {
```
No, see comment in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#discussion_r860804304

Punzo commented on src/webWorker/webWorkerManager.js:
```diff
@@ -0,0 +1,319 @@
+    }
+  }
+
+  // if no available web workers and we haven't started max web workers, start a new one
+  if (webWorkers.length < config.maxWebWorkers) {
+    spawnWebWorker()
+  }
```
You are right, they were not terminated, I have added the call in In https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/465b884827b84044f6e05551d9cd1c904e160561 :  now `VolumeImageViewer.cleanup()` runs the `terminateAllWebWorkers` function.

Regarding the `spawnWebWorker`. I advise to leave as it is for both perfomances and memory reasons. Currently the WebWorkers are created only if needed dynamically. 

Allocating all of them at viewer `constructor` or `render`, it would imply to give a N parameter for the number of workers and allocating all them at once. This was my first approach, but during testing I experienced:
1) that this can generate out of memory issue (i.e. with N ~>15). 
2) each worker would need to load the dynamic webpack bundle index.worker.js (each worker is a separate process), which is few hundreds of KB. (NOTE: each worker would also need to load the wasm, but the wasm is loaded actaully the first time `decodeAsync` is called and not at a startup).

hackermd replied:
Response to https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#issuecomment-1126027888:

Thanks for the explanation! That has improved my understanding a lot.

It's probably cleanest to instantiate and terminate web workers for every viewer instance. This may incur on overhead, but decoupling management of the web worker from viewer instances would probably be much more messy and result in more error-prone applications.

Punzo commented:
> @Punzo this is awesome! Thanks so much.
> 
> I haven't tested the code yet, but have a few general comments and suggestions. I really like the fact that you have factored out and abstracted the individual decoders and the transformer. I think there would be opportunities to further abstract it and define a constant interface for decoders and transformers by deriving them from a common abstract base class (see individual comments for details).
> 
> In addition, I am wondering whether we could improve the API for the web workers and expose some functionality at the level of the main viewer class. For example, it would be nice if the creation and termination of worker processes/threads could be controlled (and potentially parametrized) via the `render()` and `cleanup()` methods.

Thanks for the review @hackermd. I have applied/responded to the comments. Please let me know when you have the time to find IDC datasets for:

- JPEGLS
- ICC color correction
- multi-channels

Punzo commented:
uhm also I am not sure why the CI is being cancelled/failing. I fixed the syntax issue. May you please have a look?

Punzo commented:
@hackermd please let me know if I need to take any other action on this PR. Thanks!

hackermd commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,142 @@
+
+/**
+ * Initialization function that loads additional web workers and initializes them
+ * @param data
```
It's currently not very clear what data is passed around via the different messages.

In general, I prefer explicitly specifying the attributes of parameter/option objects (e.g,. `{config, workerIndex}` instead of just `data`).

Punzo replied:
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/7b34d33d4353c4ad02e06479da4e37ea3f68fc10

hackermd reviewed (changes requested):
Thanks @Punzo! This looks great! The decoding and transformation logic is now nicely encapsulated and abstracted. Much cleaner than what we had before.

I have a few more minor comments and suggestions regarding code style and documentation. Many of them evolve around the distinction between private and public functions, classes, and methods. Only functions that are exported in `dicom-microscopy-viewer.js` should be considered public and those should be kept to a minimum. All other functions should be considered private. Their name should then have a leading underscore and `@private` should be included in their docstring.

In addition, I have a few mayor comments and questions:

1. Should the `webWorkerManager` be exposed as part of the public API of the library or not? In either case, we should better document the different functions (methods) of the `webWorkerManager`.
2. When should web workers be initialized and terminated? Specifically, should a new set of workers be initialized upon rendering of a viewer instance and terminated upon cleanup or should this be handled by the application outside of the scope of the viewer class. The answer to this question should probably also dictate [1].
3. No data fetching should be performed upon construction of viewer instances. This is important for our approach to unit testing but also for the application flow. All data fetching (e.g., retrieval of bulkdata referenced in the metadata such as the ICC profiles) should be deferred and only performed when the `render()` method is called.
4. The logic for the web workers is quite sophisticated. Additional documentation will help (see [1]), but I feel we also need a couple of unit tests. Those would also help demonstrating how the API of the `webWorkerManger` is supposed to be used.

Punzo commented on package.json:
```diff
@@ -17,18 +14,25 @@
     "test": "standard 'src/*.js' && jest --silent",
-    "build": "webpack --mode=development",
-    "build:prod": "webpack --mode=production",
+    "build": "npm run webpack:bundle && npm run webpack:dynamic-import",
+    "build:prod": "npm run webpack:bundle && npm run webpack:dynamic-import",
```
I have discussed this with Erik, who spent a lot of time in analyzing perfomance of the bundle, loading time etc... with the cornerstoneWADOLoader. When using webworker the dynamic import improves the loading time (now I don't have the number now, but it is substantial) and avoid some memory issues.

Punzo commented on src/decode.js:
```diff
@@ -1,222 +1,107 @@
-const decodeFrame = ({
-  decoders,
-  transformers,
+function processDecodeAndTrasformTask (
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/7b136e66b3bbfd734744388c0594b8b741dbc2f5

Punzo commented on src/decode.js:
```diff
@@ -1,222 +1,107 @@
       throw new Error(
         'The pixel bit depth ' + bitsAllocated +
-        ' is not supported by the offscreen rendering.'
+        ' is not supported by the rendering.'
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/fbd22e21a9e2f037ef815acf287d18a504764760

Punzo commented on src/dicom-microscopy-viewer.js:
```diff
@@ -192,5 +193,6 @@ export {
   utils,
-  viewer
+  viewer,
+  webWorkerManager
```
I agree, fixed in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/342f096adb8ffc1e4caa7d9b86bb334604eedca2

Punzo commented on src/webWorker/decodeAndTrasformTask.js:
```diff
@@ -0,0 +1,184 @@
+ *
+ * @param {object} - handler data
+ * @param {function} - handler done call back
+ *
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/d30a48c43175017dd8c4fd572ed8fb0ccfe85d60

Punzo commented on src/webWorker/decoders/decoderAbstract.js:
```diff
@@ -0,0 +1,31 @@
+    return Promise.resolve()
+  }
+
+  /** Decode image.
+   *
+   * @param {Uint8Array} byteArray - Image array
+   *
+   * @returns {object} decoded array and frameInfo
+   */
+  async decodeAsync (byteArray) {
```
done in https://github.com/Punzo/dicom-microscopy-viewer/commit/154b034be7b8e5ec3052b2ae36f22e634bd87c92

Punzo commented on src/webWorker/transformers/transformerAbstract.js:
```diff
@@ -0,0 +1,16 @@
+    return Promise.resolve()
+  }
+
+  async transformAsync () {
+    if (!this.codec) {
+      await this._initialize()
+    }
+  }
```
done in https://github.com/Punzo/dicom-microscopy-viewer/commit/154b034be7b8e5ec3052b2ae36f22e634bd87c92

Punzo commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,142 @@
+/*eslint-disable*/
```
because self is defined only when loading the file in the webworker object at run time

Punzo commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,142 @@
+const taskHandlers = {}
+
+// Flag to ensure web worker is only initialized once
+let initialized = false
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/02898cb1ba601581403ef2975ada08b238db78cb

Punzo commented on src/webWorker/webWorker.js:
```diff
@@ -0,0 +1,142 @@
+/**
+ * Initialization function that loads additional web workers and initializes them
+ * @param data
+ */
```
done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/7b34d33d4353c4ad02e06479da4e37ea3f68fc10

Punzo commented:
replying here to https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#discussion_r870732169 since I think the comment was edited outside the review and there is no reply option.

> When and where are the workers initialized? Where is the configuration coming from?
> 
> The question is really how and when workers are initialized and terminated. On the one hand, it would probably make sense to initialize workers in render() and terminate them in cleanup(). On the other hand, it may also not be good idea to initialize and terminate workers for each viewer instance. Applications may switch between different slides, each of which could result viewer instances being created/destroyed.
> 

Initialization: this step is indeed pretty complicated to follow : the workers are intiliazed on the first time a startTaskOnWebWorker is called for decoding a frame (so at rendering time). All the steps are:
1) the process starts when calling `_decodeAndTransformFrame` in pyramid.js (line 483) after fecthing the frame
2) this call `_processDecodeAndTrasformTask`, which add a decoding task to the `webWorkerManager` (`addTask`)
3) `startTaskOnWebWorker` is called
4) the webManager first check if there is any available worker
5) If no webworker is instantiated or they are all busy, webManager will create a new one (`spawnWebWorker`, which create a new `DataLoader`)
6) webManager will lunch the task on the chosen webworker.

Termination: they are terminated indeed in cleanup() with the call webWorkerManager.terminateAllWebWorkers(). 

I would not be possible to pass the webworkers to other viewer instance (the web workers are an array in the webWorkerManager, there this array is instantiated in the viewer instance), therefore we need to clean them.

Punzo commented on src/viewer.js:
```diff
@@ -1227,7 +1120,8 @@ class VolumeImageViewer {
           client: this[_options].client,
-          channel: opticalPathIdentifier
+          channel: opticalPathIdentifier,
+          iccProfiles: _getIccProfiles(pyramid, this[_options].client)
```
Unfortunately there are two major issue with the fetching of the icc profiles when we initialize the wasm module for the ColorTransformer. The first in terms of performances, the second is technical: 
1) if you have N web workers (for each web worker we need to load the wasm module), you would need to fecth the icc profiles N times, wasting bandwith. 
2) we can't pass objects (i.e.,`client`) to webworkers. Therefore we actaully can't fetch the icc profiles when we initialize the wasm module for the ColorTransformer.

So I moved the fetching to the constructor (so it happens only once) and then mapped this with ColorManager transfomers by SOPInstanceUID. If you have a better idea, please let me know.

hackermd replied:
Good point!

I think we should then instantiate the decoders and transformers in `render()` rather than in `constructor()`.

Punzo replied:
the decoders and transformers are already instantiated at render time (i.e. at the first time a decoding task is requested inside a web worker object, please see https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#issuecomment-1126027888). 

While currently the ICC profiles (the buffer) are fetched at constructor level and then passed to the webworkers with the async function _createTileLoadFunction, which then pass the ICC profiles buffer to the webworker task in _decodeAndTransformFrame. 

Ideally we woul pass a `DICOMwebClient` instance into the webworker and then fetch the ICC profiles when instantiating the transfomer WASM module in each worker (this would me to fetch the ICC profiles for each worker, wasting bandwidth, but it would put the ICC profiles fecthing at rendering time). However, the main problem is that we can't pass objects (i.e. a DICOMwebClient instance) to web workers. So the only way is to pass to the webworkers, it is to pass directly the already fetched ICC profiles (the buffer) to the webworkers. 

Finally, I don't see any way to fetch them in another place which is not the constructor, because the tile loader function (`_createTileLoadFunction`) is called in the constructor (and we pass the informations regarding the tiles to the webworkers by the `_createTileLoadFunction`).

hackermd replied:
We really shouldn't do any data fetching in the constructor. That is very much against the current design of the library.

> Ideally we woul pass a DICOMwebClient instance into the webworker and then fetch the ICC profiles when instantiating the transfomer WASM module in each worker (this would imply to fetch the ICC profiles for each worker, wasting bandwidth, but it would put the ICC profiles fecthing at rendering time). However, the main problem is that we can't pass objects (i.e. a DICOMwebClient instance) to web workers. So the only way to pass the ICC profiles to the webworkers, it is to pass directly the already fetched ICC profiles (the buffer) to the webworkers.

That's fine. We can keep passing the value of ICC Profile to the workers. The main question is then where and when the workers are initialized. I tend argue that this also only happen when `render()` is called.

> I don't see any way to fetch them in another place which is not the constructor, because the tile loader function (_createTileLoadFunction) is called in the constructor (and we pass the informations regarding the tiles to the webworkers by the _createTileLoadFunction).

Wouldn't it be possible to create the tile load function and set it on the OpenLayers source in the `render()` method? If necessary, we could set a dummy tile load function in the constructor and update it once `render()` is called.

Punzo replied:
you are 100% right, we can move the fetching into the render. Done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/b36f053e618a594e34da0b4c36cf6ac92dea5764

Punzo commented:
> Thanks @Punzo! This looks great! The decoding and transformation logic is now nicely encapsulated and abstracted. Much cleaner than what we had before.
> 
> I have a few more minor comments and suggestions regarding code style and documentation. Many of them evolve around the distinction between private and public functions, classes, and methods. Only functions that are exported in `dicom-microscopy-viewer.js` should be considered public and those should be kept to a minimum. All other functions should be considered private. Their name should then have a leading underscore and `@private` should be included in their docstring.
> 
> In addition, I have a few mayor comments and questions:
> 
> 1. Should the `webWorkerManager` be exposed as part of the public API of the library or not? In either case, we should better document the different functions (methods) of the `webWorkerManager`.
> 2. When should web workers be initialized and terminated? Specifically, should a new set of workers be initialized upon rendering of a viewer instance and terminated upon cleanup or should this be handled by the application outside of the scope of the viewer class. The answer to this question should probably also dictate [1].
> 3. No data fetching should be performed upon construction of viewer instances. This is important for our approach to unit testing but also for the application flow. All data fetching (e.g., retrieval of bulkdata referenced in the metadata such as the ICC profiles) should be deferred and only performed when the `render()` method is called.
> 4. The logic for the web workers is quite sophisticated. Additional documentation will help (see [1]), but I feel we also need a couple of unit tests. Those would also help demonstrating how the API of the `webWorkerManger` is supposed to be used.

@hackermd thanks for the review. I have applied the minor comments. regarding to the major:

1) No webWorkerManager should not be in the public API. wrapper libraries should not touch the web workers. I removed the export
2) I tried to explain the process in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#issuecomment-1126027888, please let me know if now it is clear.
3) here it is the reason why I moved the fetching of the icc profiles in the constructor: https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#discussion_r872388910 . If you have a better idea, please let me know.
4) It is not possible to impelment a simple unit test for the webWorkers. We could add an integration test, but we would need to host a dataset in the CI enviroment, e.g.: https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/test/integration_test.js

hackermd commented:
> It is not possible to implement a simple unit test for the webWorkers. We could add an integration test, but we would need to host a dataset in the CI enviroment

I agree that testing the web worker logic will be hard. I was thinking to test (some of the function) that handle the messages for communication with work workers, but mock the web worker objects. For example, we could assert that the `task` array contains the expected items, etc. Would that be feasible?

Punzo commented:
> > It is not possible to implement a simple unit test for the webWorkers. We could add an integration test, but we would need to host a dataset in the CI enviroment
> 
> I agree that testing the web worker logic will be hard. I was thinking to test (some of the function) that handle the messages for communication with work workers, but mock the web worker objects. For example, we could assert that the `task` array contains the expected items, etc. Would that be feasible?

Unfortunately it is no possible.it looks like that in the jest enviroment, it is not possible to create a web worker object (i.e. when calling `spawnWebWorker` function). So it is not possible to test if the object has been created, check the array length, etc...

Punzo commented:
> 2. when fetching the ICC profiles. I think we can only in the constructor. See [ENH: decoding and ICC color transform on webworker #76 (comment)](https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76#discussion_r873406429)

done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/b36f053e618a594e34da0b4c36cf6ac92dea5764, moved the fetching in render time

Punzo commented:
@hackermd I have added some general docs description for the volume image viewer infrastructure in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/619c61dacbc8eca18d1cf23bbc542910ec0a07e1. Please let me know if it is good enough

hackermd commented:
> @hackermd I have added some general docs description for the volume image viewer infrastructure in [619c61d](https://github.com/herrmannlab/dicom-microscopy-viewer/commit/619c61dacbc8eca18d1cf23bbc542910ec0a07e1). Please let me know if it is good enough

It appears you made the changes to the HTML document. Is that correct?

These changes will get overwritten when the `npm run generateDocs` is run again. The HTML documents are autogenerated from the JavaScript source code using [JSDoc](https://jsdoc.app/). Could you please add the documentation to the docstring of the JavaScript class ([here](https://github.com/Punzo/dicom-microscopy-viewer/blob/webWorkerDecoding/src/viewer.js#L678-L684))?

Punzo commented:
> > @hackermd I have added some general docs description for the volume image viewer infrastructure in [619c61d](https://github.com/herrmannlab/dicom-microscopy-viewer/commit/619c61dacbc8eca18d1cf23bbc542910ec0a07e1). Please let me know if it is good enough
> 
> It appears you made the changes to the HTML document. Is that correct?
> 
> These changes will get overwritten when the `npm run generateDocs` is run again. The HTML documents are autogenerated from the JavaScript source code using [JSDoc](https://jsdoc.app/). Could you please add the documentation to the docstring of the JavaScript class ([here](https://github.com/Punzo/dicom-microscopy-viewer/blob/webWorkerDecoding/src/viewer.js#L678-L684))?

done in https://github.com/herrmannlab/dicom-microscopy-viewer/pull/76/commits/c87924966115cb42c7071df566824a4fda891fa2

Punzo commented:
> > @Punzo I am getting the following error upon tile loading: `TypeError: _index_worker_js__WEBPACK_IMPORTED_MODULE_0__.default is not a constructor`
> > Did you encounter this problem?
> 
> No, uhm. These are steps that I used for testing:
> 
> in dicom-microscopy-viewer:
> 
> 1. yarn install
> 2. yarn build
> 3. yarn link
> 
> in slim:
> 
> 1. yarn install
> 2. yarn link 'dicom-microscopy-viewer'
> 3. yarn start

ah I understood, you need also this one https://github.com/herrmannlab/slim/pull/93 to work on Slim

Punzo commented:
> @Punzo update of the slim build configuration fixed the problem. However, I am now running into another issue: `TypeError: can't define property "iccProfile": Object is not extensible` ([here](https://github.com/Punzo/dicom-microscopy-viewer/blob/d30a48c43175017dd8c4fd572ed8fb0ccfe85d60/src/pyramid.js#L37)).
> 
> You cannot set attributes on the metadata objects. They are immutable (see [here](https://github.com/herrmannlab/dicom-microscopy-viewer/blob/509557f8c50ee23d42acaab3e1ebe9d52c7952c9/src/metadata.js#L431)).

ok, I did not experienced this error, I will check it out asap

hackermd commented:
> > @Punzo update of the slim build configuration fixed the problem. However, I am now running into another issue: `TypeError: can't define property "iccProfile": Object is not extensible` ([here](https://github.com/Punzo/dicom-microscopy-viewer/blob/d30a48c43175017dd8c4fd572ed8fb0ccfe85d60/src/pyramid.js#L37)).
> > You cannot set attributes on the metadata objects. They are immutable (see [here](https://github.com/herrmannlab/dicom-microscopy-viewer/blob/509557f8c50ee23d42acaab3e1ebe9d52c7952c9/src/metadata.js#L431)).
> 
> ok, I did not experienced this error, I will check it out asap

Are you developing with a local DCM4CHEE backend or the Google Healthcare API? The latter doesn't handle ICC profiles yet so that code may have never been executed.

Punzo commented:
> > > @Punzo update of the slim build configuration fixed the problem. However, I am now running into another issue: `TypeError: can't define property "iccProfile": Object is not extensible` ([here](https://github.com/Punzo/dicom-microscopy-viewer/blob/d30a48c43175017dd8c4fd572ed8fb0ccfe85d60/src/pyramid.js#L37)).
> > > You cannot set attributes on the metadata objects. They are immutable (see [here](https://github.com/herrmannlab/dicom-microscopy-viewer/blob/509557f8c50ee23d42acaab3e1ebe9d52c7952c9/src/metadata.js#L431)).
> > 
> > 
> > ok, I did not experienced this error, I will check it out asap
> 
> Are you developing with a local DCM4CHEE backend or the Google Healthcare API? The latter doesn't handle ICC profiles yet so that code may have never been executed.

Ah you are right. Probably never got executed, because i have tested only on this IDC portal datasets, which have no ICC profiles:

JPEGBaseline8bit: 2.25.211094631316408413440371843585977094852/series/1.3.6.1.4.1.5962.99.1.208792987.352384958.1640886332827.2.0
JPEG2000: 2.25.106035919049347205763690043295164538046/series/1.3.6.1.4.1.5962.99.1.151372831.860614623.1640828912671.2.0

Can you share with me data with ICC profiles, otherwise I don't know how to reproduce and fix it

hackermd commented:
> Can you share with me data with ICC profiles, otherwise I don't know how to reproduce and fix it

The data sets contain ICC profiles, the Google Healthcare API just doesn't include references to them in the metadata and Slim can consequently not retrieve them. The missing BulkDataURI entries is a known bug of the Healthcare API.

As a workaround, you can download the entire data sets over DICOMweb (e.g., using the [Python dicomweb-client](https://github.com/herrmannlab/dicomweb-client)) and store them in a local instance of DCM4CHEE for development.

Punzo commented:
overdensed by https://github.com/herrmannlab/dicom-microscopy-viewer/pull/78, closing
